service/history/timerQueueAckMgr.go

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+)
+
+var (
+	timerQueueAckMgrMaxTimestamp = time.Unix(0, math.MaxInt64)
+)
+
+type (
+	timerSequenceIDs []TimerSequenceID
+
+	timerQueueAckMgrImpl struct {
+		shard         ShardContext
+		executionMgr  persistence.ExecutionManager
+		logger        bark.Logger
+		metricsClient metrics.Client
+		lastUpdated   time.Time
+		config        *Config
+
+		sync.Mutex
+		// outstanding tasks, key cluster name, value map of outstanding timer task -> finished (true)
+		clusterOutstandingTasks map[string]map[TimerSequenceID]bool
```

Comment:
I think we should keep this component simple by not making it aware of all clusters.  We should have a separate instance of TimerQueueProcessor with its own AckManager per cluster.